File: benchmark/mnist_inference_vs_jax.jl
#!/usr/bin/env julia
# Compiled PC INFERENCE (relaxation loop only, no weight-gradient training) vs upstream's own
# jax.jit -- Julia side. Companion to benchmark/mnist_inference_vs_jax.py (run THAT first --
# it generates the shared weights/batch via jax.random + KaimingInitializer, RNG-trap style,
# dumps them as raw Float32 .bin files + a shapes.txt sidecar, and produces jax_result.json +
# the converged jax.jit z_latents for the numerics cross-check below).
#
# SCOPE (docs/AUDIT_REGISTER.md section 5 J-01/J-02/J-04, docs/decisions.md section 22): this
# benchmarks Layer 2 (Reactant/XLA compiled `compile_inference`, the E-step latent relaxation
# loop only) vs Layer 0 (eager Dict `run_inference`) IN-LANGUAGE, and cross-checks the Layer 2
# result against upstream's jax.jit(run_inference) result dumped by the .py companion. This is
# NOT the J-01/J-02 weight-gradient-training compile (still blocked) -- inference only, params
# fixed throughout.
#
# Run (needs Reactant -- an env with FabricPC dev-linked + Reactant, e.g. benchmark/jit):
#   julia --project=benchmark/jit benchmark/mnist_inference_vs_jax.jl [data_dir]
# (or drive it through a warm Reactant session -- see tools/warm_session_reactant.jl -- to
# avoid repeated multi-minute Reactant cold-compiles while iterating).

using FabricPC
using FabricPC: run_inference, initialize_graph_state, compile_inference
using Random, Printf
using Reactant   # triggers FabricPCReactantExt

const BDIR = length(ARGS) >= 1 ? ARGS[1] : joinpath(@__DIR__, "mnist_inference_vs_jax_data")

function load_shapes(dir)
    d = Dict{String, Vector{Int}}()
    for line in eachline(joinpath(dir, "shapes.txt"))
        parts = split(line)
        d[parts[1]] = parse.(Int, parts[2:end])
    end
    return d
end

# Data was dumped by numpy's `ndarray.tofile()`, which ALWAYS writes C-order (row-major)
# bytes regardless of the source array's memory layout (see mnist_inference_vs_jax.py's
# dump() docstring -- an earlier version of this pair assumed `asfortranarray(...).tofile()`
# produced column-major bytes for a cheap `reshape`; it did not, silently scrambling every
# non-1xN array on read). Reconstruct correctly: reshape into the REVERSED shape (row-major
# bytes read naturally as column-major with axes reversed), then permute the axes back.
function load_arr(dir, name, shapes)
    shp = Tuple(shapes[name])
    bytes = read(joinpath(dir, "$name.bin"))
    data = reinterpret(Float32, bytes)
    nd = length(shp)
    return permutedims(reshape(collect(data), reverse(shp)), nd:-1:1)
end

function bench(f, n)
    for _ in 1:5
        f()
    end
    best = Inf
    for _ in 1:n
        best = min(best, @elapsed f())
    end
    return best
end

function main()
    shapes = load_shapes(BDIR)
    Wxh = load_arr(BDIR, "Wxh", shapes)
    bh = load_arr(BDIR, "bh", shapes)
    Why = load_arr(BDIR, "Why", shapes)
    by = load_arr(BDIR, "by", shapes)
    Xb = load_arr(BDIR, "Xb", shapes)
    Yb = load_arr(BDIR, "Yb", shapes)
    jax_jit_x = load_arr(BDIR, "jax_jit_x_z_latent", shapes)
    jax_jit_h = load_arr(BDIR, "jax_jit_h_z_latent", shapes)
    jax_jit_y = load_arr(BDIR, "jax_jit_y_z_latent", shapes)

    B, DX, DH, DY = 256, 784, 128, 10
    edge_xh = "x->h:in"
    edge_hy = "h->y:in"

    # Same NODE-CLASS CHOICE as scripts/generate_tier_d_mnist_fixtures.py / test_tier_d_mnist.jl:
    # Julia's fused `Linear` for all three nodes (== upstream's plain Linear for "x",
    # LinearExplicitGrad for "h"/"y", per Tier B/C/D).
    x = Linear((DX,), "x"; use_bias=false)
    h = Linear((DH,), "h"; activation=TanhActivation())
    y = Linear((DY,), "y"; energy=GaussianEnergy())

    inference = InferenceSGD(eta_infer=0.1, infer_steps=20, latent_decay=0.0)
    structure = graph([x, h, y], [Edge(x, h), Edge(h, y)], TaskMap(; x=x, y=y), inference)

    # RNG TRAP: GraphParams/clamps built DIRECTLY from the .py-generated arrays, never via
    # Julia's own initialize_params/RNG.
    params = GraphParams(
        Dict(
            "x" => NodeParams(
                Dict{String, Matrix{Float32}}(), Dict{String, Matrix{Float32}}()
            ),
            "h" => NodeParams(Dict(edge_xh => Wxh), Dict("b" => bh)),
            "y" => NodeParams(Dict(edge_hy => Why), Dict("b" => by))
        )
    )
    clamps = Dict{String, Any}("x" => Xb, "y" => Yb)
    rng = MersenneTwister(0)
    init = initialize_graph_state(structure, B, rng; clamps=clamps, params=params)

    # Layer 0 (eager Dict) -- correctness oracle + in-language baseline.
    eager = run_inference(params, init, clamps, structure)

    # Layer 2 (Reactant/XLA compiled) -- compile time measured separately from execution.
    println("compiling inference with Reactant/XLA ...")
    t_compile = @elapsed (ci = compile_inference(structure, params, clamps; batch=B))
    zl = ci(init)  # first post-compile execution (fast path: compile-time params reused)

    names = ci.plan.names
    idx = Dict(n => i for (i, n) in enumerate(names))
    max_eager_vs_jit = maximum(
        maximum(abs.(zl[i] .- eager.nodes[names[i]].z_latent)) for i in eachindex(zl)
    )

    # Cross-LANGUAGE numerics: Julia's Reactant-compiled converged z_latent vs upstream's
    # jax.jit converged z_latent, SAME weights/batch. A benchmark on a broken/diverged
    # computation is worthless -- this must be tight (float32 precision, ~1e-5-1e-6) before
    # the timing numbers below mean anything.
    d_x = maximum(abs.(zl[idx["x"]] .- jax_jit_x))
    d_h = maximum(abs.(zl[idx["h"]] .- jax_jit_h))
    d_y = maximum(abs.(zl[idx["y"]] .- jax_jit_y))

    t_eager = bench(() -> run_inference(params, init, clamps, structure), 10)

    # THREE JIT timings, because "julia_jit_steady_ms" was measuring something the JAX side never
    # pays. J-04's headline (7.87x) came from `ci(params, init)` — the SLOW path, which
    # `to_rarray`s every weight on EVERY call (~400 KB host->device for this MLP). JAX's timed
    # loop passes `params`/`init_state` that are ALREADY device-resident `jax.Array`s, so passing
    # them to a jitted fn copies nothing. We were charging Julia for a copy JAX doesn't make —
    # and one our own API already avoids (32b64a6 marshals params once at compile time; the
    # benchmark just never switched to it). Report all three rather than pick one, so the
    # marshalling cost is a MEASURED number instead of an attribution:
    #   remarshal — old headline: params re-marshalled per call (what 7.87x was measured on)
    #   fast      — params device-resident (compile-time), per-batch state marshalled per call.
    #               This is what a real E-step does (weights constant, batch arrives on host).
    #   resident  — params AND state device-resident: JAX's exact setup, isolating compute.
    # Residual asymmetry, stated not hidden: all three still copy the OUTPUT device->host via
    # `Array(...)`, which JAX's `block_until_ready` does not. That read is also what makes the
    # timing honest (decisions.md §11: timing the thunk without it shows a bogus ~1000x), so it
    # stays — it means these numbers are, if anything, conservative for Julia.
    t_jit_remarshal = bench(() -> ci(params, init), 30)
    t_jit = bench(() -> ci(init), 30)
    zl_host = ntuple(i -> init.nodes[ci.plan.names[i]].z_latent, length(ci.plan.names))
    zl_r = Reactant.to_rarray(zl_host)
    t_jit_resident = bench(() -> (o = ci.thunk(ci.params_r, zl_r); [Array(x) for x in o]), 30)
    # sync=true: Reactant's DOCUMENTED completion barrier (CompileOptions.jl: "blocking till the
    # computation is complete. This is recommended when benchmarking"). This is the true analog of
    # JAX's block_until_ready — it waits WITHOUT the device->host copy. We previously hand-rolled
    # the barrier as `Array(...)` and then documented the resulting copy as an unavoidable
    # asymmetry vs JAX; it was not unavoidable, it was us not reading the docs (audit 2026-07-16).
    # GATED on upstream #2647 (OPEN, "Buffer synchronisation is sometimes removed"): the emitted
    # wait can be DCE'd, which would silently restore the async artifact with no Array() left to
    # expose it. So we ASSERT the barrier is real below rather than trusting the flag.
    ci_sync = compile_inference(structure, params, clamps; batch=B, sync=true)
    t_sync = bench(() -> ci_sync.thunk(ci_sync.params_r, zl_r), 30)
    t_async_bare = bench(() -> ci.thunk(ci.params_r, zl_r), 30)
    if t_sync < 4 * t_async_bare
        @warn "#2647: sync=true's wait looks DCE'd — sync ($(t_sync*1e3) ms) is not meaningfully                slower than the bare async dispatch ($(t_async_bare*1e3) ms). Do NOT trust julia_jit_sync_ms."
    end


    println("=== RESULT ===")
    @printf("compile_time_s              = %.4f\n", t_compile)
    @printf("julia_eager_steady_ms       = %.4f\n", t_eager * 1e3)
    @printf("julia_jit_steady_ms         = %.4f   (params device-resident; HEADLINE)\n", t_jit * 1e3)
    @printf(
        "julia_jit_remarshal_ms      = %.4f   (params re-marshalled per call — the OLD J-04 path)\n",
        t_jit_remarshal * 1e3
    )
    @printf(
        "julia_jit_resident_ms       = %.4f   (params AND state device-resident — JAX's setup)\n",
        t_jit_resident * 1e3
    )
    @printf(
        "julia_jit_sync_ms           = %.4f   (sync=true barrier, NO host copy — the true JAX analog)\n",
        t_sync * 1e3
    )
    @printf(
        "  -> device->host copy cost          = %.4f ms  (resident - sync; the 'asymmetry' we used to concede)\n",
        (t_jit_resident - t_sync) * 1e3
    )
    @printf("  -> #2647 barrier check: sync/async_bare = %.0fx (want >>1; ~1 means the wait was DCE'd)\n",
            t_sync / t_async_bare)
    @printf(
        "  -> per-call param marshalling cost = %.4f ms  (remarshal - headline)\n",
        (t_jit_remarshal - t_jit) * 1e3
    )
    @printf(
        "  -> per-call state marshalling cost = %.4f ms  (headline - resident)\n",
        (t_jit - t_jit_resident) * 1e3
    )
    @printf("julia_jit_vs_julia_eager    = %.2fx\n", t_eager / t_jit)
    @printf("max|julia_jit - julia_eager| (in-language sanity) = %.3g\n", max_eager_vs_jit)
    @printf(
        "max|julia_jit - jax_jit| x/h/y (cross-language numerics) = %.3g / %.3g / %.3g\n",
        d_x, d_h, d_y
    )
end

main()

File: benchmark/mnist_inference_vs_jax.py
#!/usr/bin/env python3
"""
Compiled PC INFERENCE (relaxation loop only, no weight-gradient training) vs upstream's own
jax.jit — the MNIST-MLP architecture (x(784) -> h(128,tanh) -> y(10), same shapes/config as
examples/jit_inference.jl and Tier D's mnist track, docs/AUDIT_REGISTER.md section 6).

SCOPE (read docs/AUDIT_REGISTER.md section 5 J-01/J-02/J-04 + section 6 Tier D, and
docs/decisions.md sections 11/22 before touching this file): J-01/J-02 block compiling
`compute_local_weight_gradients` (the WEIGHT-gradient training step) under Reactant+Enzyme --
that is NOT what this benchmarks. This benchmarks `run_inference` (the E-step latent
relaxation loop) only: params and clamps ("x" the input, "y" the target -- train-style
clamping, matching examples/jit_inference.jl) are FIXED; only "h" relaxes. No weight update,
no optax, no gradient w.r.t. params anywhere in this file.

RNG TRAP (same discipline as Tier A-D, docs/AUDIT_REGISTER.md section 6): this script is the
SOLE source of the weights/batch -- KaimingInitializer-initialized via jax.random, dumped as
raw Fortran-order Float32 binaries (+ a plain-text .shapes sidecar) so the Julia side
(benchmark/mnist_inference_vs_jax.jl) builds its GraphParams/clamps DIRECTLY from these
concrete arrays, never calling its own initialize_params/RNG. Apples-to-apples numerically,
not just a timing comparison.

Run (pinned upstream venv, per this repo's CLAUDE.md JIT-lane discipline):
    /home/shivaji1012/dev-zone/FabricPC/.venv-fixtures/bin/python3 \
        benchmark/mnist_inference_vs_jax.py [outdir]
"""

import os
import sys
import time
import json

import numpy as np
import jax
import jax.numpy as jnp

from fabricpc.core.activations import TanhActivation
from fabricpc.core.energy import GaussianEnergy
from fabricpc.core.inference import InferenceSGD, run_inference
from fabricpc.core.initializers import KaimingInitializer
from fabricpc.core.types import GraphParams, NodeParams
from fabricpc.core.topology import Edge
from fabricpc.graph_assembly.graph_construction import TaskMap, graph
from fabricpc.graph_initialization.state_initializer import initialize_graph_state
from fabricpc.nodes.linear import Linear
from fabricpc.nodes.linear_explicit_grad import LinearExplicitGrad

OUTDIR = sys.argv[1] if len(sys.argv) > 1 else "."
os.makedirs(OUTDIR, exist_ok=True)

# ---------------------------------------------------------------------------------------
# Same architecture as examples/jit_inference.jl (already-validated Layer-2 example) and
# Tier D's mnist track: x(784) -> h(128, tanh) -> y(10, Identity+Gaussian). NODE-CLASS CHOICE
# mirrors scripts/generate_tier_d_mnist_fixtures.py exactly: "x" (in_degree=0) = plain Linear
# (use_bias=False -- LinearExplicitGrad's forward_and_latent_grads has no terminal-node
# special case, see that script's docstring); "h"/"y" = LinearExplicitGrad, matching Julia's
# fused `Linear` bit-for-bit (Tier B/C/D already established this).
# ---------------------------------------------------------------------------------------
B, DX, DH, DY = 256, 784, 128, 10
ETA_INFER = 0.1
INFER_STEPS = 20

key = jax.random.PRNGKey(7)


def split():
    global key
    key, sub = jax.random.split(key)
    return sub


x_node = Linear(shape=(DX,), name="x", use_bias=False)
h_node = LinearExplicitGrad(shape=(DH,), name="h", activation=TanhActivation())
y_node = LinearExplicitGrad(shape=(DY,), name="y", energy=GaussianEnergy())

inference = InferenceSGD(eta_infer=ETA_INFER, infer_steps=INFER_STEPS, latent_decay=0.0)
structure = graph(
    nodes=[x_node, h_node, y_node],
    edges=[Edge(x_node, h_node), Edge(h_node, y_node)],
    task_map=TaskMap(x=x_node, y=y_node),
    inference=inference,
)

edge_xh = "x->h:in"
edge_hy = "h->y:in"

Wxh = KaimingInitializer.initialize(split(), (DX, DH))
bh = jnp.zeros((1, DH))
Why = KaimingInitializer.initialize(split(), (DH, DY))
by = jnp.zeros((1, DY))

params = GraphParams(
    nodes={
        "x": NodeParams(weights={}, biases={}),
        "h": NodeParams(weights={edge_xh: Wxh}, biases={"b": bh}),
        "y": NodeParams(weights={edge_hy: Why}, biases={"b": by}),
    }
)

Xb = jax.random.normal(split(), (B, DX))
Yb = jax.random.normal(split(), (B, DY))
clamps = {"x": Xb, "y": Yb}

init_state = initialize_graph_state(
    structure, B, split(), clamps=clamps, params=params
)


# ---------------------------------------------------------------------------------------
# Dump shared inputs for the Julia side (RNG trap): raw C-order (row-major) Float32 binaries
# + a plain-text shapes sidecar (avoids depending on NPZ.jl in the benchmark/jit Julia env).
# NOTE: `ndarray.tofile()` ALWAYS serializes in C order regardless of the array's actual
# memory layout (confirmed empirically -- `np.asfortranarray(a).tofile()` writes IDENTICAL
# bytes to `a.tofile()`; numpy's own docs say as much). An earlier version of this script
# called `np.asfortranarray(...).tofile()` believing it produced column-major bytes for a
# cheap Julia `reshape` -- it did not, silently scrambling every non-1xN array (Wxh/Why/Xb/Yb)
# on the Julia read side (row-major bytes read back as column-major). The companion .jl loader
# now explicitly un-transposes (`permutedims(reshape(data, reverse(shape)), ndims:-1:1)`) to
# correctly invert a genuine C-order dump instead. Caught by comparing a live in-memory JAX
# array against its own disk round-trip (see benchmark git history / session notes) --
# textbook lesson: verify a serialization round-trip numerically, don't assume tofile()
# respects `asfortranarray`.
# ---------------------------------------------------------------------------------------
def dump(name, arr):
    a = np.asarray(arr, dtype=np.float32)
    a.tofile(os.path.join(OUTDIR, f"{name}.bin"))
    return name, list(a.shape)


shapes = dict(
    [
        dump("Wxh", Wxh), dump("bh", bh),
        dump("Why", Why), dump("by", by),
        dump("Xb", Xb), dump("Yb", Yb),
    ]
)
with open(os.path.join(OUTDIR, "shapes.txt"), "w") as f:
    for name, shp in shapes.items():
        f.write(f"{name} {' '.join(str(s) for s in shp)}\n")

# ---------------------------------------------------------------------------------------
# Eager reference (correctness oracle, not timed for the headline number).
# ---------------------------------------------------------------------------------------
eager_final = run_inference(params, init_state, clamps, structure)
eager_final = jax.block_until_ready(eager_final)

# ---------------------------------------------------------------------------------------
# jax.jit-compiled inference. Closes over `clamps`/`structure` (Python objects), matching
# upstream's OWN idiom for this exact call (fabricpc/training/train.py:352,
# `step_fn = jax.jit(lambda p, o, b, k: train_step(p, o, b, structure, optimizer, k))` --
# structure/optimizer closed over, never passed as a traced pytree arg) -- not a benchmark
# invention.
# ---------------------------------------------------------------------------------------
jit_run = jax.jit(lambda p, s: run_inference(p, s, clamps, structure))

t_compile0 = time.perf_counter()
jit_final = jax.block_until_ready(jit_run(params, init_state))
t_compile1 = time.perf_counter()
compile_time_s = t_compile1 - t_compile0

# Numerics check: JIT must match eager (both computing the SAME thing) before timing means
# anything.
max_abs_diff = 0.0
for name in ("x", "h", "y"):
    d = float(jnp.max(jnp.abs(jit_final.nodes[name].z_latent - eager_final.nodes[name].z_latent)))
    max_abs_diff = max(max_abs_diff, d)

# Steady-state (warm/compiled) wall-clock: min-of-N, block_until_ready each call (avoids the
# async-dispatch trap documented in decisions.md section 11 -- "timing the thunk WITHOUT
# Array(...)/block_until_ready shows a bogus ~1000x").
N = 30
for _ in range(5):
    jax.block_until_ready(jit_run(params, init_state))
times = []
for _ in range(N):
    t0 = time.perf_counter()
    jax.block_until_ready(jit_run(params, init_state))
    t1 = time.perf_counter()
    times.append(t1 - t0)
jit_steady_ms = min(times) * 1e3

# Eager steady-state, for reference (NOT the headline comparison -- Julia's own eager Dict
# path is a materially different implementation; this is just upstream-internal context).
eager_times = []
for _ in range(5):
    jax.block_until_ready(run_inference(params, init_state, clamps, structure))
for _ in range(10):
    t0 = time.perf_counter()
    jax.block_until_ready(run_inference(params, init_state, clamps, structure))
    t1 = time.perf_counter()
    eager_times.append(t1 - t0)
eager_steady_ms = min(eager_times) * 1e3

# Dump JIT-path converged z_latents for the Julia-side numerics cross-check, appending their
# shapes to the sidecar (written above, before these existed).
with open(os.path.join(OUTDIR, "shapes.txt"), "a") as f:
    for name in ("x", "h", "y"):
        n1, s1 = dump(f"jax_jit_{name}_z_latent", jit_final.nodes[name].z_latent)
        f.write(f"{n1} {' '.join(str(s) for s in s1)}\n")
        n2, s2 = dump(f"jax_eager_{name}_z_latent", eager_final.nodes[name].z_latent)
        f.write(f"{n2} {' '.join(str(s) for s in s2)}\n")

result = dict(
    batch=B, dx=DX, dh=DH, dy=DY, eta_infer=ETA_INFER, infer_steps=INFER_STEPS,
    jax_version=jax.__version__, devices=[str(d) for d in jax.devices()],
    compile_time_s=compile_time_s,
    jit_steady_ms=jit_steady_ms,
    eager_steady_ms=eager_steady_ms,
    jit_vs_eager_speedup=eager_steady_ms / jit_steady_ms,
    max_abs_diff_jit_vs_eager=max_abs_diff,
)
with open(os.path.join(OUTDIR, "jax_result.json"), "w") as f:
    json.dump(result, f, indent=2)

print(json.dumps(result, indent=2))
